### PR TITLE
Fix lost board.SPI and board.I2C

### LINKED ISSUE #3581: board.SPI lost after being used by display
While playing around with creating and destroying displays to exercise the supervisor heap, I came across some strange behavior: When a `board.SPI()` object has ever been used by a `FourWire` or a `SharpMemoryFramebuffer` that has since been deinitialized, subsequent sessions can never reconstruct it, because its pins are still in use.

Steps to reproduce:

```
Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.

Press any key to enter the REPL. Use CTRL-D to reload.
Adafruit CircuitPython 6.0.0-rc.0-61-g1a677406b-dirty on 2020-10-21; CWtest with samd51G19
>>> import board, displayio
>>> displayio.FourWire(board.SPI(), command=board.RX, chip_select=board.TX)
<FourWire>
>>> displayio.release_displays()
>>> 
soft reboot

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
Hello World!

Press any key to enter the REPL. Use CTRL-D to reload.
Adafruit CircuitPython 6.0.0-rc.0-61-g1a677406b-dirty on 2020-10-21; CWtest with samd51G19
>>> import board
>>> board.SPI()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: SCK in use
>>> 
```

or

```
Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.

Press any key to enter the REPL. Use CTRL-D to reload.
Adafruit CircuitPython 6.0.0-rc.0-61-g1a677406b-dirty on 2020-10-21; CWtest with samd51G19
>>> import board, displayio, sharpdisplay
>>> sharpdisplay.SharpMemoryFramebuffer(board.SPI(), board.RX, 400, 240)
<SharpMemoryFramebuffer>
>>> displayio.release_displays()
>>> 
soft reboot

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
Hello World!

Press any key to enter the REPL. Use CTRL-D to reload.
Adafruit CircuitPython 6.0.0-rc.0-61-g1a677406b-dirty on 2020-10-21; CWtest with samd51G19
>>> import board
>>> board.SPI()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: SCK in use
>>> 
```

What happens is:

* `common_hal_displayio_fourwire_construct()` and `common_hal_sharpdisplay_framebuffer_construct()` call `common_hal_busio_spi_never_reset(self->bus)`.
* `release_displays()` doesn’t do anything with the SPI bus.
* When the session ends, `cleanup_after_vm()` → `reset_port()` → `reset_all_pins()` does not reset the pins because they were marked as `never_reset`.
* `cleanup_after_vm()` → `reset_board_busses()` sets `spi_singleton = NULL;` so that the next call to `board.SPI()` needs to recreate it, but it can’t because the pins are still in use.

I’m not sure how this is supposed to work, but there seems to be a hole in the logic somewhere. I suppose someone should call `common_hal_busio_spi_deinit()` at some point so that the pins are made resettable again? Maybe `reset_board_busses()` before it effectively leaks the resources used by the `spi_singleton`, but then `reset_port()` and `reset_board_busses()` come in the wrong order?

Some of the observations made by @hatchman in https://github.com/adafruit/circuitpython/issues/3508#issuecomment-706211348 seem related (but probably unrelated to the actual issue of #3508, so I’m opening a new one).

### PR DESCRIPTION
**Pull request by cwalther**
The **first commit** brings `board.I2C` into the same state as `board.SPI` by fixing something that was clearly wrong, but by itself caused no severe problems. It thereby _introduces_ the problem described in #3581 for `board.SPI` also for `board.I2C`. See commit message for details.

The **second commit** fixes #3581 for both SPI and I2C.

I found no reason in the code or in the Git history for `reset_port()` to come before `reset_board_busses()` in `cleanup_after_vm()`, so I reversed them.

For symmetry, since setting `spi_singleton` to non-NULL occurs together with `common_hal_busio_spi_construct()`, setting it to NULL should also occur together with `common_hal_busio_spi_deinit()` (and likewise for I2C). This makes the pins resettable and allows the swapped `reset_port()` to do its job. (The symmetry is not complete because `common_hal_busio_spi_deinit()` can also occur without setting `spi_singleton` to NULL, i.e. `spi_singleton` can end up pointing at a non-NULL deinited object. This is dealt with in the third commit, see below.)

Another desirable symmetry would be that since `common_hal_displayio_fourwire_construct()` calls `common_hal_busio_spi_never_reset()`, `common_hal_displayio_fourwire_deinit()` should also call `common_hal_busio_spi_reset()` (which doesn’t exist, but would do the reverse of `…_never_reset()`) (and likewise for `i2cdisplay`). However, this is not necessary, since the `never_reset` flags only take effect at the end of the session, and at that time everything is already handled correctly by the changes made here. There is effectively a redundancy between the `never_reset` flags and `reset_board_busses()` checking for things still in use by displays. Therefore I avoided adding that symmetry, which would just add more code.

Critical review appreciated.

The **third commit** fixes a tangentially related issue: After one had explicitly deinited the `board.SPI` or `board.I2C` busses (maybe to temporarily use their pins for other purposes), there was no way to get them back in the same session. Calling the function again would just return the unusable deinited object. Reproducible as follows:

```
Adafruit CircuitPython 6.0.0-rc.0-99-g32c2e9354-dirty on 2020-10-25; PewPew M4 CW with samd51G19
>>> import board
>>> bus = board.SPI()
>>> bus.try_lock()
True
>>> bus.write(b'X')
>>> bus.unlock()
>>> bus.deinit()
>>> 
>>> bus = board.SPI()
>>> bus.try_lock()
True
>>> bus.write(b'X')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Object has been deinitialized and can no longer be used. Create a new object.
>>> 
```

or

```
# with an SSD1306 display connected to the I2C pins
Adafruit CircuitPython 6.0.0-rc.0-99-g32c2e9354-dirty on 2020-10-25; PewPew M4 CW with samd51G19
>>> import board
>>> bus = board.I2C()
>>> bus.try_lock()
True
>>> bus.writeto(60, b'X')
>>> bus.unlock()
>>> bus.deinit()
>>> 
>>> bus = board.I2C()
>>> bus.try_lock()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Object has been deinitialized and can no longer be used. Create a new object.
>>> 
```

Testing on non-SAMD ports would be appreciated.

### CODE REVIEW

**Review by jepler — COMMENTED** [review]
This looks good to me but I would also like @tannewt to weigh in on the movement of reset_port in particular.  Thank you!

**Review by dhalbert — APPROVED** [review]
Thanks for your careful analysis!

File: main.c [hunk 1]
```diff
@@ -234,10 +234,10 @@ void cleanup_after_vm(supervisor_allocation* heap) {
     common_hal_canio_reset();
     #endif
 
-    reset_port();
     #if CIRCUITPY_BOARD
     reset_board_busses();
     #endif
+    reset_port();
```
[the remark this answers is not in the document]
Comment (cwalther) [reply main.c:242]: Good point, done.

**Review by tannewt — CHANGES_REQUESTED** [review]
This looks really good to me! I think we'll want this in 6.0.x though and then we can merge 6.0.x into `main`.

**Comment by tannewt** [comment]
Would you mind rebasing onto 6.0.x and then changing the PR's base? Thanks!

**Comment by cwalther** [comment]
Done.

**Review by tannewt — APPROVED** [review]
Thank you!

**Comment by cwalther** [comment]
Thanks!

**Pull request merged**